server/conn.go

```diff
@@ -1049,6 +1054,13 @@ func (cc *clientConn) Run(ctx context.Context) {
 	// The client connection would detect the events when it fails to change status
 	// by CAS operation, it would then take some actions accordingly.
 	for {
+		// Close connection between txn when we are going to shutdown server.
+		if cc.server.inShutdownMode.Load() {
+			if !cc.ctx.GetSessionVars().InTxn() {
+				return
+			}
+		}
+
 		if !atomic.CompareAndSwapInt32(&cc.status, connStatusDispatching, connStatusReading) ||
 			// The judge below will not be hit by all means,
 			// But keep it stayed as a reminder and for the code reference for connStatusWaitShutdown.
```

Comment:
ref https://github.com/pingcap/tidb/blob/9e086b987c8cf07ef3691a90084be692d5380888/server/server.go#L793-L812

Same as before, It may be updated as `connStatusWaitShutdown` and the CAS will fail, as the comment "keep it stayed as a reminder and for the code reference for connStatusWaitShutdown."